### ISSUE TITLE
Fusing the optimizer step into the backward pass

### ISSUE THREAD

**Issue by martin-marek** [jax-ml/jax#30338]
I am interested in fine-tuning a large model in a memory-constrained setting. In particular, I am interested in using a small batch size combined with a simple optimizer, so >90% of my memory use is just model + grads. If I could fuse the backward pass with the optimizer step, this could potentially halve the memory requirements for training / fine-tuning.

In PyTorch, it's possible to [fuse](https://docs.pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html) the optimizer step into the backward pass using `Tensor.register_post_accumulate_grad_hook(hook)`.

Is it possible to do something similar in JAX?

**Comment by mattjj** [comment]
Thanks for the question!

If the model update step is under a `jax.jit`, i.e. both the gradient computation and the application of the optimizer update, I would expect XLA to do that for you _unless_ something is defeating that optimization, e.g. you have a scan over layers in your model. Can you say more about your situation, so we can figure out why it's not being automatically applied as an XLA optimization?

To write it manually, you can use a `jax.custom_vjp`. If you could share your code, or a representative toy model of it, we can probably advise exactly where to put it, but the basic idea is you can put the optimizer update in the bwd rule. Then the output of `jax.grad` would be the updated optimizer state rather than actual gradients.

**Comment by reachtarunhere** [comment]
@mattjj  Thanks for the reply. Am I correct to understand that you mean XLA should interleave gradient computation and optimizer step? ideally once you have computed the gradient, used it for optimization and the layer before you has computed the gradient then this gradient tensor is not needed. So this should save some memory. Maybe it doesn't matter because peak memory usage is right after the forward pass due to activations ?

[PyTorch memory reduction by fusion of optimizer step in backward pass](https://docs.pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html#disclaimer-this-technique-is-not-for-alll)

**Comment by martin-marek** [comment]
Thank you @mattjj for your quick response!

You are completely right and I feel silly for not realizing this on my own. Indeed if I `jax.jit` the training step and donate the model/optimizer state (e.g. `@partial(jax.jit, donate_argnames=('optimizer_state'))`) then the optimizer update is fused to the backward pass and full model gradients are never materialized. It didn't occur to me that the jit compiler would do this by default since PyTorch requires a manual implementation.

@reachtarunhere if you use gradient checkpointing (`jax.checkpoint`) to reduce the size of activations, then the difference between materializing / not materializing gradients becomes more significant.
